src/platform/backends/shared/base_virtual_machine.cpp

```diff
@@ -258,8 +260,11 @@ std::string mp::BaseVirtualMachine::ssh_exec(const std::string& cmd, bool whispe
 
 void mp::BaseVirtualMachine::renew_ssh_session()
 {
-    if (!MP_UTILS.is_running(current_state())) // spend time updating state only if we need a new session
-        throw SSHException{fmt::format("SSH unavailable on instance {}: not running", vm_name)};
+    {
+        const std::unique_lock lock{state_mutex};
+        if (!MP_UTILS.is_running(current_state())) // spend time updating state only if we need a new session
+            throw SSHException{fmt::format("SSH unavailable on instance {}: not running", vm_name)};
+    }
```

Comment:
Changes to the `ssh_session` below are still unsafe (if not super likely to cause issues).